### ISSUE TITLE
SSH problems with .sh scripts as SSH client

### ISSUE THREAD

**Issue by tmssngr** [git-lfs/git-lfs#3064]
I'm currently playing with Git-LFS in combination with GitLab via SSH and noticed some problems (internal SmartGit 18.2 preview, Git 2.17.1, Windows 10). When having selected "Use SmartGit as SSH client" in the SmartGit preferences, SmartGit sets `GIT_SSH` to `smartssh.sh` (SmartGit 18.1 still uses `smartssh.cmd`) and `GIT_SSH_VARIANT` to `ssh` (the directory for "smartssh.sh" is prepended to PATH). Git executes this command just fine (`GIT_TRACE` set to `1`):
    
```
09:23:14.984206 run-command.c:640       trace: run_command: unset GIT_CONFIG_PARAMETERS GIT_PREFIX; smartssh.sh -p 1234 singer@smart 'git-upload-pack '\''/home/singer/arduino-projects.git'\'''
```
    
but the Git-LFS filter has some problems:
    
```
09:24:19.751993 trace git-lfs: run_command: smartssh.sh git@192.168.92.207 git-lfs-authenticate tom/test.git download
09:24:19.756875 trace git-lfs: ssh: git@192.168.92.207 failed, error: fork/exec E:\java\smartgit\master\build\dist-templates\scripts\smartssh.sh: %1 is not a valid Win32 application., message: 
```
    
Could it be that Git-LFS tries to executed the SSH script in a different environment than Git?

**Comment by ttaylorr** [comment]
Hi @tmssngr, thanks for opening this issue and sorry that you're having problems. It looks like Git LFS is executing the SSH override fine (note the `run_command:` line in the trace output).

But the SSH command itself is failing on `git@192.168.92.207` with the message:

> ```ShellSession
> fork/exec E:\java\smartgit\master\build\dist-templates\scripts\smartssh.sh: %1 is not a valid Win32 application., message: 
> ```

So, are you able to verify that that your remote is able to handle the command `git-lfs-authenticate`? In other words, are you able to run the following yourself:

```ShellSession
$ smartssh.sh git@192.168.92.207 git-lfs-authenticate tom/test.git download
```

**Comment by tmssngr** [comment]
Hi @ttaylorr,
For testing purposes I've added
`echo "$@" >> /c/temp/smartssh.txt`
to `smartssh.sh` and judging from the content of `/c/temp/smartssh.txt` Git-LFS definitely fails to execute this script (while Git succeeds). It seems Git launches it with the help of its Linux-simulating environment while Git-LFS tries to execute it as a Windows application that obviously fails to handle `.sh` files.

**Comment by ttaylorr** [comment]
Hm. This seems like a bug, then -- @larsxschneider @sschuberth do either of you have thoughts?

**Comment by sschuberth** [comment]
I wouldn't necessarily call this a bug in `git-lfs` but rather a missing feature. Windows cannot execute `.sh` files directly. In Git-for-Windows there's [some logic](https://github.com/git-for-windows/git/blob/090a092/run-command.c#L251) to leverage `sh` if the file to execute is an `.sh` script. `git-lfs` would need to do the same and prefix the command line with `sh -c` if the file to run is an `.sh` script and we're on Windows.

**Comment by ttaylorr** [comment]
@sschuberth I see, thanks. It sounds like we need to do something similar for v2.5.0 (or later). I'll queue this for that.

@tmssngr in the meantime, can you change your SSH environment variables to invoke `sh` instead?

**Comment by tmssngr** [comment]
As an intermediate workaround we've switched back to using .cmd files on Windows. We need to avoid this because it causes problems with cmd.exe that is said to handle spaces in the path differently than Git.

**Comment by sschuberth** [comment]
> cmd.exe that is said to handle spaces in the path differently than Git.

That's pretty vague (and the first time I hear this). Could you provide some details? In general, using `.cmd` on Windows sounds like the correct thing to do.

**Comment by homo-programmatis** [comment]
@tmssngr refers to an issue explained here: https://stackoverflow.com/questions/12891383

In short, cmd.exe doesn't like receiving TWO quoted parameters:

This works:
`cmd.exe /c notepad.exe "C:\1.txt"`
`cmd.exe /c "notepad.exe" C:\1.txt`

This fails with "The filename, directory name, or volume label syntax is incorrect"
`cmd.exe /c "notepad.exe" "C:\1.txt"`

You can try it yourself.

**Comment by sschuberth** [comment]
But this works: `cmd.exe /c ""notepad.exe" "C:\1.txt""`, i.e. simply but the whole string in quotes, too.

**Comment by homo-programmatis** [comment]
That's true, but then git has to recognize it's going to use cmd.exe and tailor the commandline appropriately, which it currently doesn't do.

**Comment by bk2204** [comment]
I think this may be fixed if you use `GIT_SSH_COMMAND` instead of `GIT_SSH` on `master` (as of 7990ab17a4ec9b039df61f3e08873f2c687379e5).  That isn't in a released version yet, but I'd be interested to know if it fixes the issue for you.

The difference is that `GIT_SSH_COMMAND` is passed to the shell (in git-lfs, as of the commit above) and `GIT_SSH` is not.
